[DataPipe] Adding BatcherMapDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68197

Implements part of #57031

cc @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

NivekT committed:
[DataPipe] Adding BatcherMapDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440963).

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440963).

ejguan commented on torch/utils/data/datapipes/map/grouping.py:
```diff
@@ -2,12 +2,76 @@
+                raise IndexError(f"Index {index} is out of bound.")
+
+    def __len__(self) -> int:
+        if self.length is not None:
+            return self.length
+        if isinstance(self.datapipe, Sized) and self.unbatch_level == 0:
+            if self.drop_last:
+                self.length = len(self.datapipe) // self.batch_size
+            else:
+                self.length = (len(self.datapipe) + self.batch_size - 1) // self.batch_size
+            return self.length
+        raise TypeError("{} instance doesn't have valid length".format(type(self).__name__))
```
IMO, we should assume `__len__` is always implemented for `MapDataPipe`

NivekT replied:
I agree, though there are cases where it can be hard to compute (especially when one element in the input DataPipe can expands into many).

ejguan replied:
The unbatch_level is super annoying here. Otherwise, we don't need to check if self.datapipe is Sized.

NivekT replied:
I think you still need to check for it before you call `len(self.datapipe)` right?

In either case, maybe we should get rid of `unbatch_level`? I feel that users can call `.unbatch` prior to this if they wanted that operation. I decided to include it only because `IterDataPipe`'s version has that argument and I hesitate to make the two different.

ejguan replied:
If we assume prior datapipe has `__len__`, then we can directly call `len(self.datapipe`.

> In either case, maybe we should get rid of unbatch_level? I feel that users can call .unbatch prior to this if they wanted that operation. I decided to include it only because IterDataPipe's version has that argument and I hesitate to make the two different.

Yeah. I feel like `unbatch_level` making me confused even for IterDataPipe. At least based on my experience with Vision and Text, no one really uses it.

NivekT replied:
If that is the case, I think we should remove it for simplicity. I will make a PR and let the domains users have a quick look. I agree that it is unlikely that people are using it.

ejguan replied:
Sound great. Thanks

ejguan reviewed (approved):
LGTM, and you may need to change pyi for Dataset.

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"


Implements part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440963).

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"


Implements part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"


Implements part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440963).

NivekT committed:
Update on "[DataPipe] Adding BatcherMapDataPipe"


Implements part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440963](https://our.internmc.facebook.com/intern/diff/D32440963)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440963).
